### PR TITLE
Add somes types to warnings

### PR DESCRIPTION
**Pull request by tharvik**
Improve `warnings` a bit more, by being more precise on accepted types (such as adding `Optional` where `None` is handled).
Also, add types to `warnings.catch_warnings`.

### CODE REVIEW

File: stdlib/2and3/warnings.pyi [hunk 1]
```diff
@@ -1,33 +1,37 @@
 # Stubs for warnings
 
-# Based on http://docs.python.org/3.2/library/warnings.html
+from typing import Any, Dict, NamedTuple, Optional, TextIO, Union
+from types import ModuleType, TracebackType
 
-from typing import Any, List, TextIO, Union
-
-def warn(message: Union[str, Warning], category: type = ...,
+def warn(message: Union[str, Warning], category: Optional[type] = ...,
```
Comment (gvanrossum) [inline stdlib/2and3/warnings.pyi]: Now that Type[C] exists we can describe the type of category more precisely: it must be `Optional[Type[Warning]]`.

File: stdlib/2and3/warnings.pyi [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
-                  lineno: int, module: str = ..., registry: Any = ...,
-                  module_globals: Any = ...) -> None: ...
-
-# logging modifies showwarning => make it a variable.
-def _showwarning(message: str, category: type, filename: str, lineno: int,
-                 file: TextIO = ..., line: str = ...) -> None: ...
-showwarning = _showwarning
-
+                  lineno: int, module: Optional[str] = ...,
+                  registry: Optional[Dict[Any, Any]] = ...,
```
Comment (gvanrossum) [inline stdlib/2and3/warnings.pyi]: Looks like the registry type is

File: stdlib/2and3/warnings.pyi [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
 class catch_warnings:
-    # TODO record and module must be keyword arguments!
-    # TODO type of module?
-    def __init__(self, record: bool = ..., module: Any = ...) -> None: ...
-    def __enter__(self) -> List[Any]: ...
-    def __exit__(self, type, value, traceback) -> bool: ...
+    def __init__(self, *, record: bool = ...,
+                 module: Optional[ModuleType] = ...) -> None: ...
+    def __enter__(self) -> Optional[List[_Record]]: ...
+    def __exit__(self, exc_type: Optional[type], exc_val: Optional[Exception],
```
Comment (gvanrossum) [inline stdlib/2and3/warnings.pyi]: For `type` here we could use `Type[BaseException]`.

File: stdlib/2and3/warnings.pyi (hunk 2 above)
Comment (tharvik) [inline stdlib/2and3/warnings.pyi]: Yeay, it's not clearly documented, I don't think that `warn_explicit` should even be exposed or at least, it should be more documented.

File: stdlib/2and3/warnings.pyi (hunk 1 above)
Comment (tharvik) [inline stdlib/2and3/warnings.pyi]: Hoo, we have that now, nice!

**Pull request merged**